### PR TITLE
EDM Diffusion Models

### PR DESCRIPTION
**Pull request by mnabian**
# Modulus Pull Request

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -49,6 +49,7 @@ repos:
     hooks:
     - id: ruff
       args: [--fix]
+      exclude: 'examples'
```
Comment (dallasfoster) [inline .pre-commit-config.yaml:52]: 1. Is this exclusion of the examples folder going to be a general rule or is it because you would rather not have to lint the files in the present folder right now? 

2. Please keep exceptions and exclusions to the `pyproject.toml` file under the `[tool.ruff]` header.

File: examples/generative/diffusion/conf/config.yaml [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+
+
+# # Weather-related
+# data_config: ?          # String to include the data config
+# task: ?                 # String to include the task
+# data_type: ?            # String to include the data type
+
+# # Regression
+# ckpt_unet: ?            # Checkpoint for the UNet to predict the mean
```
Comment (dallasfoster) [inline examples/generative/diffusion/conf/config.yaml:92]: Unless you have a good reason for the inclusion of these commented lines, I would just as soon not include them in order to prevent confusion.

File: examples/generative/diffusion/fid.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    sigma -= mu.ger(mu) * len(dataset_obj)
+    sigma /= len(dataset_obj) - 1
+    return mu.cpu().numpy(), sigma.cpu().numpy()
+
+
+def calculate_fid_from_inception_stats(mu, sigma, mu_ref, sigma_ref):
+    m = np.square(mu - mu_ref).sum()
+    s, _ = scipy.linalg.sqrtm(np.dot(sigma, sigma_ref), disp=False)
+    fid = m + np.trace(sigma + sigma_ref - s * 2)
+    return float(np.real(fid))
```
Comment (dallasfoster) [inline examples/generative/diffusion/fid.py:122]: 1. This should be moved to `modulus/metrics`.
2. scipy.ilnalg.sqrtm for this matrix is not stable generally.

File: examples/generative/diffusion/generate.py [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+import torch
+import tqdm
+from omegaconf import DictConfig
+from utils import StackedRandomGenerator, open_url
+
+from modulus.distributed import DistributedManager
+from modulus.launch.logging import PythonLogger, RankZeroLoggingWrapper
+
+
+def sampler(
```
Comment (dallasfoster) [inline examples/generative/diffusion/generate.py:30]: This sampler is overly complex, poorly documented, and not modular. If we intend to support generative models like these in the future, it would behoove us to have general purpose langevin samplers.

File: modulus/metrics/diffusion/loss.py [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+    beta_d: float, optional
+        Coefficient for the diffusion process, by default 19.9.
+    beta_min: float, optional
+        Minimum bound, by defaults 0.1.
+    epsilon_t: float, optional
+        Small positive value, by default 1e-5.
+
+    Note:
+    -----
+    Reference: Song, Y., Sohl-Dickstein, J., Kingma, D.P., Kumar, A., Ermon, S. and
```
Comment (dallasfoster) [inline modulus/metrics/diffusion/loss.py:39]: I don't think it should be necessary for a user to reference the paper in order to understand the notation, terminology, and purpose of the class.  In this case, I think __call__ and sigma should have documentation as well.

File: modulus/metrics/diffusion/loss.py [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+        Maximum supported noise level, by default 100.0.
+
+    Note:
+    -----
+    Reference: Song, Y., Sohl-Dickstein, J., Kingma, D.P., Kumar, A., Ermon, S. and
+    Poole, B., 2020. Score-based generative modeling through stochastic differential
+    equations. arXiv preprint arXiv:2011.13456.
+    """
+
+    def __init__(self, sigma_min: float = 0.02, sigma_max: float = 100.0):
```
Comment (dallasfoster) [inline modulus/metrics/diffusion/loss.py:105]: Same comments as for `VPLoss`, documentation needs to be improved.

File: modulus/metrics/diffusion/loss.py [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+        y, augment_labels = (
+            augment_pipe(images) if augment_pipe is not None else (images, None)
+        )
+        n = torch.randn_like(y) * sigma
+        D_yn = net(y + n, sigma, labels, augment_labels=augment_labels)
+        loss = weight * ((D_yn - y) ** 2)
+        return loss
+
+
+# class EDMLoss:
```
Comment (dallasfoster) [inline modulus/metrics/diffusion/loss.py:122]: Why is this class commented out? We should either uncomment or not include it in the code.

File: modulus/metrics/diffusion/loss.py [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+    def __init__(
+        self, P_mean: float = 0.0, P_std: float = 1.2, sigma_data: float = 0.5
+    ):
+        self.P_mean = P_mean
+        self.P_std = P_std
+        self.sigma_data = sigma_data
+
+        with torch.no_grad():
+            # resume_state_dump = '/lustre/fsw/nvresearch/mmardani/output/logs/edm_era5_cwb_12chans_fcn_4x_crop448_zarr_fulldata_unetregression/complex-mayfly_2023.04.25_16.42/output/training-state-006924.pt'
+            resume_state_dump = "/lustre/fsw/nvresearch/mmardani/output/logs/edm_era5_cwb_12chans_fcn_4x_crop448_zarr_fulldata_unetregression/mindful-parakeet_2023.04.03_22.54/output/training-state-042650.pt"
```
Comment (dallasfoster) [inline modulus/metrics/diffusion/loss.py:344]: This is inappropriate to include in public code.

File: modulus/metrics/diffusion/loss.py
Comment (dallasfoster) [inline modulus/metrics/diffusion/loss.py:1]: All of these classes should have better documentation but also consider whether they should be classes or not. Sometimes the class structure for metrics is just duct tape for poor workflows.

File: modulus/models/diffusion/__init__.py
Comment (dallasfoster) [inline modulus/models/diffusion/__init__.py:1]: Consider whether these networks should live under `models/diffusion`, `models/` or maybe even `models/unet`.

File: modulus/models/diffusion/utils.py
Comment (dallasfoster) [inline modulus/models/diffusion/utils.py:1]: It is not clear why this is necessary or why it mixes numpy and torch.

File: modulus/models/diffusion/layers.py [hunk 9]
```diff
@@ -0,0 +1,541 @@
...
+from typing import Any, Dict, List
+
+import numpy as np
+import torch
+from torch.nn.functional import silu
+
+from modulus.models.diffusion import weight_init
+
+
+class Linear(torch.nn.Module):
```
Comment (dallasfoster) [inline modulus/models/diffusion/layers.py:29]: Is this function necessary as a standalone layer under `models/diffusion`. It seems like the only purpose for its existance is to support layer initialization. If so, then maybe we should include it under `models/layers`.

File: modulus/models/diffusion/layers.py
Comment (dallasfoster) [inline modulus/models/diffusion/layers.py:1]: Consider whether these layers should be general enough to include under `models/layers`

File: examples/generative/diffusion/dataset/__init__.py
Comment (dallasfoster) [inline examples/generative/diffusion/dataset/__init__.py:1]: The dataset function should really be reconsidered and determine whether they can be generalized and moved out of examples.

**Review by dallasfoster — COMMENTED** [review]
There are a lot of places where documentation can be improved and notation is not entirely explain. 

The examples folder should at least include a readme with an explanation about how to use the example.

The placement of the layers and folders should be reconsidered and generalized wherever possible.

Creation of a generic, modular, langevin diffusion class should be considered as a separate epic.

There are several places with large chunks of commented out code, please consider uncommenting or removing these pieces of code.

Please let me know which of these and the other comments you think should be in-scope for this PR.

File: examples/generative/diffusion/generate.py (hunk 4 above)
Comment (mnabian) [reply examples/generative/diffusion/generate.py:30]: Added to this epic: https://github.com/NVIDIA/modulus/issues/196

File: modulus/models/diffusion/__init__.py
Comment (mnabian) [reply modulus/models/diffusion/__init__.py:1]: Added to this epic: https://github.com/NVIDIA/modulus/issues/196

File: modulus/models/diffusion/layers.py (hunk 9 above)
Comment (mnabian) [reply modulus/models/diffusion/layers.py:29]: Added to this epic: https://github.com/NVIDIA/modulus/issues/196

File: modulus/models/diffusion/layers.py
Comment (mnabian) [reply modulus/models/diffusion/layers.py:1]: Added to this epic: https://github.com/NVIDIA/modulus/issues/196

File: modulus/metrics/diffusion/loss.py (hunk 7 above)
Comment (mnabian) [reply modulus/metrics/diffusion/loss.py:167]: The `EDMLoss` class is WIP, to generalize it to different models including ResDiff. the commented out code helps to guide this generalization.

File: modulus/metrics/diffusion/loss.py
Comment (mnabian) [reply modulus/metrics/diffusion/loss.py:1]: Improved the documentation. Added the class vs function issue to this epic: https://github.com/NVIDIA/modulus/issues/196

**Comment by mnabian** [comment]
Thanks @dallasfoster for reviewing this PR!

> There are a lot of places where documentation can be improved and notation is not entirely explain.
I agree. I added more documentation to `loss.py`, and I will keep improving the documentation in the subsequent PRs.

> The examples folder should at least include a readme with an explanation about how to use the example.
The example is a work in progress. Once the example is ready, I will also include a README with detailed instructions.

> The placement of the layers and folders should be reconsidered and generalized wherever possible.
I agree. I have added this to the list of issues in this epic: https://github.com/NVIDIA/modulus/issues/196. Will decide about this once I develop the example and get a better understanding of what can be generalized.

> Creation of a generic, modular, langevin diffusion class should be considered as a separate epic.
For now I added it to this epic: https://github.com/NVIDIA/modulus/issues/196

> There are several places with large chunks of commented out code, please consider uncommenting or removing these pieces of code.
The commented out code is mostly the ResDiff code and it helps me generalize the existing code to add support for ResDiff and super-resolution task. I propose we keep them in this PR, and I can remove the remaining comments right before the release.

File: modulus/metrics/diffusion/loss.py (hunk 5 above)
Comment (mnabian) [reply modulus/metrics/diffusion/loss.py:39]: Added docstrings for call. Sigma already has docstrings

File: modulus/models/diffusion/utils.py
Comment (mnabian) [reply modulus/models/diffusion/utils.py:1]: Added to this epic: https://github.com/NVIDIA/modulus/issues/196
I think this was mainly because they did not want to have an `nn` module as the input to their model, as that can potentially cause issues with pickling. I'll try to figure this out and most likely we should be able to get rid of this entirely.

File: test/metrics/diffusion/test_losses.py [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+
+    # With augmentation
+    def mock_augment_pipe(imgs):
+        return imgs, None
+
+    loss_value_with_augmentation = loss_func(fake_net, img, labels, mock_augment_pipe)
+    assert isinstance(loss_value_with_augmentation, torch.Tensor)
+
+
+# RegressionLoss tests
```
Comment (loliverhennigh) [inline test/metrics/diffusion/test_losses.py:142]: Maybe delete this?

**Review by loliverhennigh — APPROVED** [review]
Hey @mnabian, to be honest I struggled a bit with giving this a review. I ran through all the code. There is quite a lot of detail with how then diffusion models work and I am not very familiar with all the variations in this space. Certainty a lot more complex then just basic regression like many of our examples.

At this point I don't really have any good advice about integrating this better with modulus other then some of the comments Dallas made such as moving over some of the layer functions over. I think we should continue to treat this as a long term project following your epic, https://github.com/NVIDIA/modulus/issues/196. Lets just get this in and as the examples build I think we will be able to work out abstractions that make sense...

File: modulus/metrics/diffusion/loss.py (hunk 6 above)
Comment (mnabian) [reply modulus/metrics/diffusion/loss.py:142]: Added documentation for the `__call__` method

File: test/metrics/diffusion/test_losses.py (hunk 10 above)
Comment (mnabian) [reply test/metrics/diffusion/test_losses.py:142]: Thanks Oliver for your comment. Those tests are work in progress and I should be able to complete and uncomment those in a subsequent PR.

**Review by dallasfoster — APPROVED** [review]
Thank you for adding the necessary tasks to the epic list. I also appreciate the additional documentation. I think at some point we would want to include the examples in the ruff linting, but I understand that might be a separate PR after the launch examples have been moved. Everything else here looks manageable.
